dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubSourceCodeRepo.java

```diff
@@ -1055,27 +1055,88 @@ public String getRepositoryId(Entry entry) {
         }
     }
 
+
+
     @Override
     public String getMainBranch(Entry entry, String repositoryId) {
         String mainBranch = null;
 
         // Get repository based on username and repo id
+        mainBranch = getDefaultBranch(repositoryId);
+        if (mainBranch == null) {
+            return null;
+        }
+        // Determine which branch to use for tool info
+        if (entry.getDefaultVersion() != null) {
+            mainBranch = getBranchNameFromDefaultVersion(entry);
+        }
+
+        return mainBranch;
+    }
+
+    @Override
+    public String getDefaultBranch(String repositoryId) {
         if (repositoryId != null) {
             try {
                 GHRepository repository = github.getRepository(repositoryId);
-                // Determine the default branch on Github
-                mainBranch = repository.getDefaultBranch();
+                // Determine the default branch on GitHub
+                return repository.getDefaultBranch();
             } catch (IOException e) {
                 LOG.error("Unable to retrieve default branch for repository " + repositoryId, e);
                 return null;
             }
         }
-        // Determine which branch to use for tool info
-        if (entry.getDefaultVersion() != null) {
-            mainBranch = getBranchNameFromDefaultVersion(entry);
-        }
+        return null;
+    }
 
-        return mainBranch;
+    /**
+     * Detect branches that include a .dockstore.yml using a variety of heuristics.
+     * <p>
+     * Heuristics include looking at the default branch, commits that just touched a .dockstore.yml at the end of a branch, etc. while trying to avoid too many
+     * API calls, particularly an unbounded number of calls based on factors that can change between repos (number of tags, branches, etc.)
+     *
+     * @param repositoryId
+     * @return
+     */
+    public Set<String> detectDockstoreYml(String repositoryId) {
+        Set<String> candidateBranches = new HashSet<>();
+        if (repositoryId != null) {
+            try {
+                GHRepository repository = getRepository(repositoryId);
+                // see if there is a .dockstore.yml on any path that was just added to a branch/the HEAD
+                // unfortunately, there is no nice way to map this to a specific tag or branch, arbitrarily pick 5 to try
+                final int arbitraryNumberOfGuesses = 5;
+                final List<GHCommit> ghCommits = repository.queryCommits().path(DOCKSTORE_YML_PATH).pageSize(arbitraryNumberOfGuesses).list().toList();
```

Comment:
There's a plural `DOCKSTORE_YML_PATHS` constant, because the file can be in two locations, root and `.github`. :( At a glance, our doc seems to emphasize the root directory more, but...

I suggest a comment if we're going to limit this to the root.

Reply:
Adding check (but lowering number of guesses)